### PR TITLE
build binary releases with LDC

### PR DESCRIPTION
**Pull request by MartinNowak**
- curl no longer needed for dlang.org as docs are build on linux
- use current libcurl (the one to be released)
- drop custom binaries being copied from previous release
- use ldc to build release
- setup VC environment to build create_dmd_release on Windows
- update to debian 9 (stretch)
- add changelog entry

### CODE REVIEW

**Comment by MartinNowak** [comment]
FYI @kinke and @JohanEngelen

**Comment by kinke** [comment]
> rmdir /s /q ..\generated
> rmdir: failed to remove '/s': No such file or directory
> [...]
> mingw32-make: *** [win32.mak:140: clean] Error 1

This is probably because DM make isn't used anymore, and the MinGW stuff preferring the (Posix) tools in its dir over the cmd.exe builtins. I don't think you can get around DM make here.

File: create_dmd_release/build_all.d [hunk 1]
```diff
@@ -269,63 +259,48 @@ auto addPrefix(R)(R rng, string prefix)
...
-
-            if (ldcVer.empty)
-                dmd = `old-dmd\dmd2\windows\bin\dmd.exe`;
-            else
-                dmd = `ldc\ldc2-`~ldcVer~`-windows-multilib\bin\ldmd2.exe`;
-            rdmd = `old-dmd\dmd2\windows\bin\rdmd.exe`;
+            // copy libcurl needed for create_dmd_release
+            cmd(`copy extraBins\dmd2\windows\bin\libcurl.dll .`);
+            // setup VC to build create_dmd_release (https://stackoverflow.com/a/2124759/2371032)
+            cmd(`cmd /c "`~'`'~`"$($env:VSINSTALLDIR)\VC\Auxiliary\Build\vcvarsall.bat`~'`'~`" x64 && set" |
```
[the remark this answers is not in the document]
Comment (MartinNowak) [reply create_dmd_release/build_all.d:284]: I can try to get rid of that, my problem was mostly that an optlink link.exe was hiding the MSVC linker (likely from a DMC installation). Maybe I can just remove that now.
Didn't knew about auto-detection.

build_release.bat is just used by the azure CI test, the releases are build on VMs directly by build_all.d.

**Comment by MartinNowak** [comment]
- [ ] try auto-detection of MSVC (hide or uninstall dm)
- [ ] fix make on Azure/Windows
- [ ] fix changelog build -> built

**Comment by kinke** [comment]
Trying to help out, I've tried this monster of a process natively on Win64 with VS 2019. It was a terrible experience, and after countless failed attempts, now at least the generation of the windows.7z archive seems to work. The biggest problem of course is the mix of a DMC toolchain for the 32-bit OMF parts (incl. dmc.exe, lib.exe and implib.exe - lib.exe and link.exe colliding with the MSVC names...), a 32-bit MSVC toolchain for the 32-bit COFF parts, and finally a 64-bit MSVC toolchain for the 64-bit parts, and everything being built at once to generate one fat package.

Here's what I did:

1) Patch this PR a bit:
```diff
diff --git a/create_dmd_release/build_all.d b/create_dmd_release/build_all.d
index 4b63e67..3d1eaa2 100644
--- a/create_dmd_release/build_all.d
+++ b/create_dmd_release/build_all.d
@@ -279,16 +279,7 @@ void runBuild(ref Box box, string ver, bool isBranch, bool skipDocs, string ldcV
             break;
         case OS.windows:
             // copy libcurl needed for create_dmd_release
-            cmd(`copy extraBins\dmd2\windows\bin\libcurl.dll .`);
-            // setup VC to build create_dmd_release (https://stackoverflow.com/a/2124759/2371032)
-            cmd(`cmd /c "`~'`'~`"$($env:VSINSTALLDIR)\VC\Auxiliary\Build\vcvarsall.bat`~'`'~`" x64 && set" |
-              foreach {
-                if ($_ -match "=") {
-                  $v = $_.split("="); set-item -force -path "ENV:\$($v[0])" -value "$($v[1])"
-                }
-              }
-            `);
-
+            cmd("copy ldc/ldc2-"~ldcVer~"-windows-multilib/bin/libcurl.dll .");
             dmd = "ldc/ldc2-"~ldcVer~"-windows-multilib/bin/ldmd2.exe";
             rdmd = "ldc/ldc2-"~ldcVer~"-windows-multilib/bin/rdmd.exe --compiler="~dmd;
             break;
@@ -538,7 +529,7 @@ int main(string[] args)
         getCodesignCerts(workDir~"/codesign");
     foreach (url; ldcCompilers.map!(s =>
             "https://github.com/ldc-developers/ldc/releases/download/v"~ldcVer~"/"~s))
-        fetchFile(url, cacheDir~"/"~baseName(url), !verifySignature);
+        fetchFile(url, cacheDir~"/"~baseName(url));
     fetchFile("http://ftp.digitalmars.com/"~optlink, cacheDir~"/"~optlink);
     fetchFile("http://ftp.digitalmars.com/"~libC, cacheDir~"/"~libC);
     fetchFile("http://downloads.dlang.org/other/"~libCurl, cacheDir~"/"~libCurl, verifySignature);
diff --git a/create_dmd_release/common.d b/create_dmd_release/common.d
index 8a94ea9..19944c3 100644
--- a/create_dmd_release/common.d
+++ b/create_dmd_release/common.d
@@ -287,7 +287,7 @@ void archiveLZMA(string inputDir, string archive)
     scope(exit) chdir(saveDir);
     chdir(dirName(inputDir));

-    auto cmd = archive.endsWith(".7z") ? ["7za", "a", archive, baseName(inputDir)] :
+    auto cmd = archive.endsWith(".7z") ? ["7z", "a", archive, baseName(inputDir)] :
             ["tar", "-Jcf", archive, baseName(inputDir)];
     auto rc = execute(cmd);
     enforce(!rc.status, rc.output);
diff --git a/create_dmd_release/create_dmd_release.d b/create_dmd_release/create_dmd_release.d
index 394b1c2..4970807 100644
--- a/create_dmd_release/create_dmd_release.d
+++ b/create_dmd_release/create_dmd_release.d
@@ -361,17 +361,15 @@ void buildAll(Bits bits, string branch, bool dmdOnly=false)
     auto msvcVarsX86 = "";
     auto msvcVars = "";
     auto msvcEnv = "";
-    version(Windows)
+    version(Windows) if (bits == Bits.bits64)
     {
         // Just overwrite any logic in makefiles and leave setup to vcvarsall.bat
         msvcEnv = ` "VCDIR=" "SDKDIR=" "CC=cl" "CC32=cl" "LD=link" "AR=lib"`;

-        // ensure LDC auto-configures proper environment
-        auto ldcVars = hostIsLDC ? "set LDC_VSDIR_FORCE=1 && " : null;
         // Setup MSVC environment for x64/x86 native builds
-        auto vcVars = quote(environment["VSINSTALLDIR"] ~ `VC\Auxiliary\Build\vcvarsall.bat`);
-        msvcVarsX64 = vcVars~" x64 && " ~ ldcVars;
-        msvcVarsX86 = vcVars~" x86 && " ~ ldcVars;
+        auto vcVars = quote(environment["CREATE_DMD_VSINSTALLDIR"] ~ `VC\Auxiliary\Build\vcvarsall.bat`);
+        msvcVarsX64 = vcVars~" x64 && ";
+        msvcVarsX86 = vcVars~" x86 && ";
         msvcVars = bits == Bits.bits64 ? msvcVarsX64 : msvcVarsX86;
     }

```
2) Prerequisites: 
   - Some D host compiler.
   - Download & extract DMC; I've also had to download the **implib.exe** tool from a different DM package, and extracted it to the DMC bin dir. Considering what this tool already does, this seems like a step that should be included. [DMC also already includes required DM make.exe.]
   - 7-Zip.
   - git.
3) Start a naked `cmd.exe` shell; my environment is pretty clean, no D compilers and no VS etc. set up.
4) `set PATH=C:\LDC\dmc\bin;%PATH%;C:\Program Files\7-Zip;C:\Program Files\Git\bin`
5) `set CREATE_DMD_VSINSTALLDIR=C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\`
6) Build the `build_all` tool: `dmd -g build_all.d common.d -version=NoVagrant`
  Then copy the `libcurl.dll` of matching bitness to that dir so that `build_all.exe` can load it.
7) `build_all.exe v1.23.0 master --skip-docs --skip-verify`

**Comment by kinke** [comment]
The Windows issues should be fixed by the extra commit in #468. No idea about the Vagrant (?) build.

**Comment by MartinNowak** [comment]
>     4\. I've renamed the previously used `VSINSTALLDIR` var, because that var has a special meaning for LDC [it assumes a fully set up MSVC environment then].

What's missing from the vcvarsall setup?
https://github.com/ldc-developers/ldc/blob/3f1a3e683a93e402997c21cbfd92fc1f2039ee89/driver/tool.cpp#L219-L226

It seems we still need to call vsvarsall to setup cl for C files, no?

For now I'd rather stick with the VSINSTALLDIR + LDC_VSDIR_FORCE combination to avoid changing the VM.

**Comment by MartinNowak** [comment]
> The Windows issues should be fixed by the extra commit in #468. No idea about the Vagrant (?) build.

Thanks a lot, I've integrated most your changes.

**Comment by rainers** [comment]
> rmdir /s /q ..\generated
> rmdir: failed to remove '/s': No such file or directory

To disambiguate, using the traditional `rd` might help. But instead of using `make clean`, wouldn't some sort of `git clean` be better? I suspect the platform dependent `clean` in the make files leaves files behind, causing the Windows version to have `-dirty` suffix, see https://issues.dlang.org/show_bug.cgi?id=20620

**Comment by MartinNowak** [comment]
> To disambiguate, using the traditional `rd` might help. But instead of using `make clean`, wouldn't some sort of `git clean` be better? I suspect the platform dependent `clean` in the make files leaves files behind, causing the Windows version to have `-dirty` suffix, see https://issues.dlang.org/show_bug.cgi?id=20620

I agree, but I don't want to get into fixing other repos (or replacing dm make), so we just adjust paths accordingly for now.
Feel invited to switch over to build.d if that one is ready enough for all repos, I can deal with any VM woes accordingly to the Windows CI test.

**Comment by kinke** [comment]
> What's missing from the vcvarsall setup?
> https://github.com/ldc-developers/ldc/blob/3f1a3e683a93e402997c21cbfd92fc1f2039ee89/driver/tool.cpp#L219-L226

Right now with the only partially adopted changes, VSINSTALLDIR is only required for these tools here, and only to find the right batch file for setting up the environment, which is then only required for win64.mak (but also set up for win32.mak, where it's superfluous, LDC_VSDIR_FORCE alone would suffice and should already be set according to https://github.com/dlang/installer/blob/720d3af0756390c065774b62952076fc73d17a49/create_dmd_release/build_all.d#L285). If it's run inside a naked environment with only that var set, not a full vcvarsall.bat env, this requires that LDC_VSDIR_FORCE kludge, which, after a quick glance, should fortunately still yield a consistent environment (it usually detects the latest MSVC toolchain). I still think this is suboptimal, and renaming the var would be much cleaner & clearer. LDC_VSDIR_FORCE was introduced because of Rainer's troubles with this repo here, and renaming VSINSTALLDIR would have sufficed to overcome this problem.

**Comment by kinke** [comment]
> but also set up for win32.mak, where it's superfluous

Hmm nope, that makes sure the 32-bit DMD is linked against the same MSVC runtime as the 64-bit one, so it's not superfluous. [Setting `LDC_VSDIR` instead would have the same effect, but using that instead of `VSINSTALLDIR` for the vcvarsall.bat lookup is probably not self-descriptive.]

**Comment by rainers** [comment]
> I agree, but I don't want to get into fixing other repos (or replacing dm make), so we just adjust paths accordingly for now.

I didn't mean to change the make files, but to run `git clean` instead. See https://github.com/dlang/installer/pull/469

**Pull request merged**